File: app/models/shipment.rb
class Shipment < ApplicationRecord
  has_snowflake_id('sr')
  acts_as_paranoid
  audit_trail only: [ :checked_in_at, :lab_id, :label ]

  belongs_to :organization
  belongs_to :user
  belongs_to :lab
  belongs_to :container_transfer
  has_many :containers
  has_many :implementation_shipment_items

  scope :pending_check_in, -> { where(checked_in_at: nil) }

  validates_presence_of :organization_id
  validates_inclusion_of :shipment_type, in: [ 'sample', 'implementation' ]

  before_create :add_shipping_label

  SHIPMENT_CODE_CHARS = ('A'..'Z').to_a

  def self.generate_uniq_shipment_codes(num)
    ::StringUtil.generate_uniq_random_strings(SHIPMENT_CODE_CHARS, 3, num)
  end

  SHORT_JSON = {
    only: [ :shipment_type, :created_at, :checked_in_at, :created_by, :id, :label, :editable,
           :data, :name, :organization_id, :packing_url, :user_id ],
    include: {
      organization: {
        only: [ :id, :name ],
        include: {},
        methods: []
      },
      user: {
        only: [ :id, :first_name, :last_name ],
        include: {},
        methods: [ :profile_img_url, :name ]
      }
    },
    methods: [ :status ]
  }.freeze

  INTERNAL_SHIPMENT_CHECKIN_JSON = {
    only: [ :shipment_type, :created_at, :checked_in_at, :created_by, :id, :label, :editable,
           :data, :name, :contact_name, :contact_number, :packing_url,
           :scheduled_pickup, :note, :receiving_note, :pickup_street, :pickup_zipcode, :organization_id ],
    include: {
      organization: {
        only: [ :id, :name, :test_account ],
        include: {},
        methods: []
      },
      containers: {
        only: [ :id, :label, :barcode, :container_type_id, :shipment_code,
               :storage_condition, :slot, :row, :status, :location_id, :cover ],
        include: {
          container_type: {
            only: [ :id, :name, :shortname, :acceptable_lids ],
            include: {},
            methods: []
          },
          location: {
            only: [ :id, :name ],
            include: {},
            methods: []
          }
        },
        methods: [ :suggested_user_barcode ]
      }
    },
    methods: [ :status ]
  }.freeze

  FULL_JSON = {
    only: [ :shipment_type, :created_at, :checked_in_at, :created_by, :id, :label, :editable,
           :data, :name, :contact_name, :contact_number, :packing_url,
           :scheduled_pickup, :note, :receiving_note, :pickup_street, :pickup_zipcode, :organization_id, :lab_id ],
    include: {
      organization: {
        only: [ :id, :name, :test_account ],
        include: {},
        methods: []
      },
      containers: {}
    },
    methods: [ :status ]
  }.freeze

  LIST_JSON = Shipment::FULL_JSON.merge({
    include: {
      organization: {
        only: [ :id, :name, :test_account ],
        include: {},
        methods: []
      },
      containers: {
        include: {}
      }
    }
  }).freeze

  def status
    if checked_in_at.nil?
      'pending'
    else
      'received'
    end
  end

  def checkedin_containers
    containers.where(status: "available")
  end

  def inbound_containers
    containers.where(status: "inbound")
  end

  def all_containers_checkedin?
    not inbound_containers.any?
  end

  def serializable_hash(opts = {})
    opts = SHORT_JSON.merge(opts || {})
    super(opts)
  end

  def checkin!(current_admin)
    self.checked_in_at = Time.now
    self.checked_in_by = current_admin.id
    self.editable = false

    self.save!

    if self.shipment_type == 'implementation'
      ImplementationShipmentSlackMessage.perform_async(self.id)
    elsif self.shipment_type == 'sample'
      NOTIFICATION_SERVICE.shipment_checked_in(self)
    end
  end

  private

  def add_shipping_label
    l = nil
    loop do
      l = make_shipping_label
      break if Shipment.where(label: l, checked_in_at: nil).empty?
    end
    self.label = l
  end

  ShippingLabelChars = 'ABCDEFGHJKMNPQRSTUVWXYZ'.split(//)
  def make_shipping_label
    ShippingLabelChars.sample(4).join
  end
end

File: app/client/main/models/Shipment/Shipment.js
import PropTypes from 'prop-types';

export const shipmentPropTypes = PropTypes.shape({
  id: PropTypes.func.isRequired,
  type: PropTypes.func.isRequired,
  name: PropTypes.func,
  containerIds: PropTypes.func,
  label: PropTypes.func,
  organizationName: PropTypes.func.isRequired,
  createdAt: PropTypes.func,
  checkedInAt: PropTypes.func,
  isCheckedIn: PropTypes.func,
  isContainerCheckedIn: PropTypes.func,
  isDeletable: PropTypes.func,
  status: PropTypes.func,
  isPickup: PropTypes.func,
  contactName: PropTypes.func,
  contactNumber: PropTypes.func,
  scheduledPickup: PropTypes.func,
  note: PropTypes.func,
  receiving_note: PropTypes.func,
  pickupStreet: PropTypes.func,
  pickupZipcode: PropTypes.func,
  packingUrl: PropTypes.func,
  labId: PropTypes.func,
  fullShipment: PropTypes.func,
});

class ShipmentModel {
  constructor(_shipment) {
    this._shipment = _shipment;
  }

  id() { return this._shipment.get('id'); }

  type() { return this._shipment.get('shipment_type'); }

  name() { return this._shipment.get('name'); }

  containerIds() { return this._shipment.get('container_ids', []); }

  label() { return this._shipment.get('label'); }

  organizationName() { return this._shipment.getIn(['organization', 'name']); }

  createdAt() { return this._shipment.get('created_at'); }

  checkedInAt() {
    return this._shipment.get('checked_in_at');
  }

  isCheckedIn() {
    return this.checkedInAt() != undefined;
  }

  static isContainerCheckedIn(container) {
    return container.get('status') !== 'inbound';
  }

  isDeletable() {
    return !this.isCheckedIn();
  }

  status() {
    const status = this._shipment.get('status');
    if (status) { return status.replace(/_/g, ' '); }
    return undefined;
  }

  isPickup() { return (this.scheduledPickup() !== undefined); }
  contactName() { return this._shipment.get('contact_name'); }
  contactNumber() { return this._shipment.get('contact_number'); }
  scheduledPickup() { return this._shipment.get('scheduled_pickup'); }
  note() { return this._shipment.get('note'); }
  receiving_note() { return this._shipment.get('receiving_note'); }
  pickupStreet() { return this._shipment.get('pickup_street'); }
  pickupZipcode() { return this._shipment.get('pickup_zipcode'); }
  packingUrl() { return this._shipment.get('packing_url'); }
  labId() { return this._shipment.get('lab_id'); }

  fullShipment() { return this._shipment; }
}

export default ShipmentModel;
